sdk/lib/opentelemetry/sdk/trace/span.rb

```diff
@@ -7,9 +7,241 @@
 module OpenTelemetry
   module SDK
     module Trace
+      # Implementation of {OpenTelemetry::Trace::Span} that records trace events.
+      #
+      # rubocop:disable Metrics/ClassLength
       class Span < OpenTelemetry::Trace::Span
-        # TODO
+        # TODO: does this need synchronization? I don't think so...
+        attr_reader :name
+
+        # Return the flag whether this span is recording events
+        #
+        # @return [Boolean] true if this Span is active and recording information
+        #   like events with the #add_event operation and attributes using
+        #   #set_attribute.
+        def recording_events?
+          true
+        end
+
+        # Set attribute
+        #
+        # Note that the OpenTelemetry project
+        # TODO: This doc link is broken
+        # {https://github.com/open-telemetry/opentelemetry-specification/blob/master/semantic-conventions.md
+        # documents} certain "standard attributes" that have prescribed semantic
+        # meanings.
+        #
+        # @param [String] key
+        # @param [String, Boolean, Numeric] value
+        #
+        # @return [self] returns itself
+        def set_attribute(key, value)
+          super
+          @mutex.synchronize do
+            if @ended
+              OpenTelemetry.logger.warn('Calling set_attribute on an ended Span.')
+            else
+              @attributes ||= {}
+              @attributes[key] = value
+              @trace_config.trim_attributes(@attributes, :max_attributes_count)
+              @total_recorded_attributes += 1
+            end
+          end
+          self
+        end
+
+        # Add an Event to Span
+        #
+        # Note that the OpenTelemetry project
+        # {https://github.com/open-telemetry/opentelemetry-specification/blob/master/semantic-conventions.md
+        # documents} certain "standard event names and keys" which have
+        # prescribed semantic meanings.
+        #
+        # @param [String, Callable] name_or_event_formatter The name of the event
+        #   or an EventFormatter, a lazily evaluated callable that returns an
+        #   Event instance.
+        # @param [optional Hash<String, Object>] attrs One or more key:value pairs, where
+        #   the keys must be strings and the values may be string, boolean or
+        #   numeric type. This argument should only be used when passing in a
+        #   name, not an EventFormatter.
+        # @param [optional Time] timestamp Timestamp for the event.
+        #
+        # @return [self] returns itself
+        def add_event(name_or_event_formatter, attrs = nil, timestamp: nil)
```

Comment:
@sveiss and I found that this has some unfortunate side effects (further context here)
https://bugs.ruby-lang.org/issues/12717 

One suggestion might be changing these 3 to all be pos arguments
```ruby
def add_event(name_or_event_formatter, attrs = nil, timestamp = nil)
```

Reply:
@ibawt and I had concerns about the ergonomics of this API anyway. If we leave it as is, I'd prefer to go the other way with keyword args for both optionals:
```ruby
def add_event(name_or_event_formatter, attributes: nil, timestamp: nil)
```
However, the "lazy" form of this requires a lambda instead of the more natural (IMO) block. I'd prefer either splitting the API into `add_event(name, attributes: nil, timestamp: nil)` and `add_lazy_event(&block)` **or** taking an implicit block and making all other args optional:
```ruby
def add_event(name: nil, attributes: nil, timestamp: nil)
  raise ArgumentError unless block_given? == (name.nil? && attributes.nil? && timestamp.nil?)
  raise ArgumentError unless Internal.valid_attributes?(attrs)
  # ^ in the API, v in the SDK
  event = block_given? ? yield : Event.new(name: name, attributes: attributes, timestamp: timestamp || Time.now)
  @mutex.synchronize do ... end
end
```

Reply:
I went with the second approach.